### PR TITLE
Implement FixedRateOfForeignAsset

### PR DESCRIPTION
**Pull request by zjb0807**
Closes: #1621

### CODE REVIEW

File: modules/asset-registry/src/lib.rs [hunk 1]
```diff
@@ -275,6 +280,88 @@ impl<T: Config> ForeignAssetIdMapping<ForeignAssetId, MultiLocation, AssetMetada
 	}
 }
 
+pub struct FixedRateOfForeignAsset<T, F: Get<u128>, R: TakeRevenue>(
+	Weight,
+	u128,
+	Option<MultiLocation>,
+	PhantomData<(T, F, R)>,
+);
```
Comment (shaunxw) [inline modules/asset-registry/src/lib.rs:288]: Use fields for readability, and we need more expressive generic params naming.

File: modules/asset-registry/src/lib.rs [hunk 2]
```diff
@@ -275,6 +280,88 @@ impl<T: Config> ForeignAssetIdMapping<ForeignAssetId, MultiLocation, AssetMetada
 	}
 }
 
+pub struct FixedRateOfForeignAsset<T, F: Get<u128>, R: TakeRevenue>(
```
Comment (shaunxw) [inline modules/asset-registry/src/lib.rs:283]: Need docstring.

File: modules/asset-registry/src/lib.rs [hunk 3]
```diff
@@ -275,6 +280,88 @@ impl<T: Config> ForeignAssetIdMapping<ForeignAssetId, MultiLocation, AssetMetada
...
+
+impl<T: Config, F: Get<u128>, R: TakeRevenue> WeightTrader for FixedRateOfForeignAsset<T, F, R> {
+	fn new() -> Self {
+		Self(0, 0, None, PhantomData)
+	}
+
+	fn buy_weight(&mut self, weight: Weight, payment: Assets) -> Result<Assets, XcmError> {
+		log::trace!(target: "asset-registry::weight", "buy_weight weight: {:?}, payment: {:?}", weight, payment);
+
+		for (index, multi_asset) in payment.fungible.iter().enumerate() {
```
Comment (shaunxw) [inline modules/asset-registry/src/lib.rs:298]: You can use `fungible.iter().next()` to get the first element.

File: modules/asset-registry/src/lib.rs [hunk 4]
```diff
@@ -275,6 +280,88 @@ impl<T: Config> ForeignAssetIdMapping<ForeignAssetId, MultiLocation, AssetMetada
...
+						return Ok(unused);
+					}
+				}
+				AssetId::Abstract(_) => {
+					// do nothing
+				}
+			}
+		}
+
+		log::trace!(target: "asset-registry::weight", "Did not match");
```
Comment (shaunxw) [inline modules/asset-registry/src/lib.rs:335]:
```suggestion
		log::trace!(target: "asset-registry::weight", "no concrete fungible asset");
```

File: modules/asset-registry/src/lib.rs [hunk 5]
```diff
@@ -275,6 +280,88 @@ impl<T: Config> ForeignAssetIdMapping<ForeignAssetId, MultiLocation, AssetMetada
...
+		for (index, multi_asset) in payment.fungible.iter().enumerate() {
+			// TODO: only support first fungible assets now.
+			// Maybe be removed in the future.
+			if index > 0 {
+				log::trace!(target: "asset-registry::weight", "buy_weight only support one fungible assets");
+				return Err(XcmError::TooExpensive);
+			}
+
+			match multi_asset.0 {
+				AssetId::Concrete(ref multi_location) => {
```
Comment (shaunxw) [inline modules/asset-registry/src/lib.rs:307]: `if let` would be more idiomatic

File: modules/asset-registry/src/lib.rs [hunk 6]
```diff
@@ -275,6 +280,88 @@ impl<T: Config> ForeignAssetIdMapping<ForeignAssetId, MultiLocation, AssetMetada
...
+					log::debug!(target: "asset-registry::weight", "buy_weight multi_location: {:?}", multi_location);
+					if let Some(CurrencyId::ForeignAsset(_)) =
+						Pallet::<T>::location_to_currency_ids(multi_location.clone())
+					{
+						let amount = F::get() * (weight as u128) / (WEIGHT_PER_SECOND as u128);
```
Comment (shaunxw) [inline modules/asset-registry/src/lib.rs:312]: use safe math

File: modules/asset-registry/src/lib.rs [hunk 7]
```diff
@@ -275,6 +280,88 @@ impl<T: Config> ForeignAssetIdMapping<ForeignAssetId, MultiLocation, AssetMetada
...
+		}
+
+		log::trace!(target: "asset-registry::weight", "Did not match");
+		Err(XcmError::TooExpensive)
+	}
+
+	fn refund_weight(&mut self, weight: Weight) -> Option<MultiAsset> {
+		log::trace!(target: "asset-registry::weight", "refund_weight weight: {:?}, weight: {:?}, amount: {:?}, multi_location: {:?}", weight, self.0, self.1, self.2);
+		let weight = weight.min(self.0);
+		let amount = F::get() * (weight as u128) / (WEIGHT_PER_SECOND as u128);
```
Comment (shaunxw) [inline modules/asset-registry/src/lib.rs:342]: Use same math.

File: modules/asset-registry/src/lib.rs [hunk 8]
```diff
@@ -275,6 +280,88 @@ impl<T: Config> ForeignAssetIdMapping<ForeignAssetId, MultiLocation, AssetMetada
...
+
+		self.0 = self.0.saturating_sub(weight);
+		self.1 = self.1.saturating_sub(amount);
+
+		log::trace!(target: "asset-registry::weight", "refund_weight amount: {:?}", amount);
+		if amount > 0 && self.2.is_some() {
+			Some((self.2.as_ref().expect("checked; qed").clone(), amount).into())
```
Comment (shaunxw) [inline modules/asset-registry/src/lib.rs:349]: `checked; qed` doesn't contain any information

File: modules/asset-registry/src/lib.rs [hunk 9]
```diff
@@ -275,6 +280,88 @@ impl<T: Config> ForeignAssetIdMapping<ForeignAssetId, MultiLocation, AssetMetada
...
+			None
+		}
+	}
+}
+
+impl<T, F: Get<u128>, R: TakeRevenue> Drop for FixedRateOfForeignAsset<T, F, R> {
+	fn drop(&mut self) {
+		log::trace!(target: "asset-registry::weight", "take revenue, weight: {:?}, amount: {:?}, multi_location: {:?}", self.0, self.1, self.2);
+		if self.1 > 0 && self.2.is_some() {
+			R::take_revenue((self.2.as_ref().expect("checked; qed").clone(), self.1).into());
```
Comment (shaunxw) [inline modules/asset-registry/src/lib.rs:360]: same here

File: modules/asset-registry/src/lib.rs (hunk 3 above)
Comment (zjb0807) [reply modules/asset-registry/src/lib.rs:298]: Will there be multiple Assets here? I want to support all, but for the time being only support the first one

Comment (shaunxw) [reply modules/asset-registry/src/lib.rs]: We can always update the impl if multiple assets support is needed in the future.

File: modules/asset-registry/src/lib.rs (hunk 1 above)
Comment (shaunxw) [reply modules/asset-registry/src/lib.rs]: tuple struct is not readable with 3 elements like `self.0`, `self.1`, `self.2`

File: modules/asset-registry/src/lib.rs [hunk 10]
```diff
@@ -275,6 +280,105 @@ impl<T: Config> ForeignAssetIdMapping<ForeignAssetId, MultiLocation, AssetMetada
 	}
 }
 
+/// Simple fee calculator that requires payment in a single fungible at a fixed rate.
+///
+/// The constant `FixedRate` type parameter should be the concrete fungible ID and the amount of it
+/// required for one second of weight.
+pub struct FixedRateOfForeignAsset<T, FixedRate: Get<u128>, R: TakeRevenue> {
```
Comment (shaunxw) [inline modules/asset-registry/src/lib.rs:287]: Is the generic param `T` used somewhere?

File: runtime/integration-tests/src/relaychain/kusama_cross_chain_transfer.rs [hunk 11]
```diff
@@ -154,3 +154,155 @@ fn transfer_to_sibling() {
 		assert_eq!(Tokens::free_balance(BNC, &AccountId::from(ALICE)), 94_989_760_000_000);
 	});
 }
+
+#[test]
+fn test_asset_registry_module() {
+	TestNet::reset();
+
+	fn sibling_2000_account() -> AccountId {
```
Comment (shaunxw) [inline runtime/integration-tests/src/relaychain/kusama_cross_chain_transfer.rs:162]:
```suggestion
	fn karura_reserve_account() -> AccountId {
```

**Review by shaunxw — APPROVED** [review]
LGTM. Some final nitpicks.

File: modules/asset-registry/src/lib.rs (hunk 10 above)
Comment (zjb0807) [reply modules/asset-registry/src/lib.rs:287]: https://github.com/AcalaNetwork/Acala/blob/21dfd9edca4012dd730147ca32227fb019709531/modules/asset-registry/src/lib.rs#L316

used here.

Comment (xlc) [reply modules/asset-registry/src/lib.rs:287]: the rate will be different for each asset. it should read minimal balance of the asset and use ratio between native token ED to determine the weight per second

**Review by xlc — APPROVED** [review]

File: modules/asset-registry/src/lib.rs [hunk 12]
```diff
@@ -275,6 +283,132 @@ impl<T: Config> ForeignAssetIdMapping<ForeignAssetId, MultiLocation, AssetMetada
...
+					let ed = T::Currency::minimum_balance();
+
+					log::trace!(
+						target: "asset-registry::weight",
+						"buy_weight FixedRate: {:?}, asset minimal_balance: {:?}, native token ed: {:?}",
+						FixedRate::get(), asset_metadatas.minimal_balance, ed
+					);
+
+					ensure!(!ed.is_zero(), XcmError::TooExpensive);
+					ensure!(!WEIGHT_PER_SECOND.is_zero(), XcmError::TooExpensive);
```
[the remark this answers is not in the document]
Comment (zjb0807) [reply modules/asset-registry/src/lib.rs:338]: Ok. I'll remove the `ed` too. The integration tests can ensure it is non-zero.

**Review by xlc — DISMISSED** [review]

**Review by xlc — APPROVED** [review]

**Pull request merged**